lib/libc/uuid/uuid.3
.\"	$NetBSD: uuid.3,v 1.5.16.1 2008/05/18 12:30:21 yamt Exp $
.\"
.\" Copyright (c) 2004 The NetBSD Foundation, Inc.
.\" All rights reserved.
.\"
.\" This code is derived from software contributed to The NetBSD Foundation
.\" by Jason R. Thorpe.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE NETBSD FOUNDATION, INC. AND CONTRIBUTORS
.\" ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
.\" TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
.\" PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE FOUNDATION OR CONTRIBUTORS
.\" BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
.\" CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
.\" SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
.\" INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
.\" CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
.\" ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
.\" POSSIBILITY OF SUCH DAMAGE.
.\"
.\" Copyright (c) 2002 Marcel Moolenaar
.\" Copyright (c) 2002 Hiten Mahesh Pandya
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
.\" IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
.\" OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
.\" IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
.\" INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
.\" BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
.\" LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
.\" AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
.\" OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\" $FreeBSD: src/lib/libc/uuid/uuid.3,v 1.4 2003/08/08 19:12:28 marcel Exp $
.\"
.Dd April 22, 2008
.Dt UUID 3
.Os
.Sh NAME
.Nm uuid_compare , uuid_create , uuid_create_nil , uuid_equal ,
.Nm uuid_from_string , uuid_hash , uuid_is_nil , uuid_to_string ,
.Nm uuid_enc_le , uuid_dec_le , uuid_enc_be , uuid_dec_be
.Nd Universally Unique Identifier routines
.Sh LIBRARY
.Lb libc
.Sh SYNOPSIS
.In uuid.h
.Ft int32_t
.Fn uuid_compare "const uuid_t *uuid1" "const uuid_t *uuid2" "uint32_t *status"
.Ft void
.Fn uuid_create "uuid_t *uuid" "uint32_t *status"
.Ft void
.Fn uuid_create_nil "uuid_t *uuid" "uint32_t *status"
.Ft int32_t
.Fn uuid_equal "const uuid_t *uuid1" "const uuid_t *uuid2" "uint32_t *status"
.Ft void
.Fn uuid_from_string "const char *str" "uuid_t *uuid" "uint32_t *status"
.Ft uint16_t
.Fn uuid_hash "const uuid_t *uuid" "uint32_t *status"
.Ft int32_t
.Fn uuid_is_nil "const uuid_t *uuid" "uint32_t *status"
.Ft void
.Fn uuid_to_string "const uuid_t *uuid" "char **str" "uint32_t *status"
.Ft void
.Fn uuid_enc_le "void *buf" "const uuid_t *uuid"
.Ft void
.Fn uuid_dec_le "const void *buf" "uuid_t *"
.Ft void
.Fn uuid_enc_be "void *buf" "const uuid_t *uuid"
.Ft void
.Fn uuid_dec_be "const void *buf" "uuid_t *"
.Sh DESCRIPTION
These routines provide for the creation and manipulation of Universally
Unique Identifiers
.Pq UUIDs ,
also referred to as Globally Unique Identifiers
.Pq GUIDs .
.Pp
The
.Fn uuid_compare
function compares two UUIDs.
It returns \-1 if
.Fa uuid1
precedes
.Fa uuid2 ,
0 if they are equal, or 1 if
.Fa uuid1
follows
.Fa uuid2 .
.Pp
The
.Fn uuid_create
function creates a new UUID.
Storage for the new UUID must be pre-allocated by the caller.
.Pp
The
.Fn uuid_create_nil
function creates a nil-valued UUID.
Storage for the new UUID must be pre-allocated by the caller.
.Pp
The
.Fn uuid_equal
function compares two UUIDs to determine if they are equal.
It returns 1 if they are equal, and 0 if they are not equal.
.Pp
The
.Fn uuid_from_string
function parses a 36-character string representation of a UUID and
converts it to binary representation.
Storage for the UUID must be pre-allocated by the caller.
.Pp
The
.Fn uuid_hash
function generates a hash value for the specified UUID.
Note that the hash value is not a cryptographic hash, and should not be
assumed to be unique given two different UUIDs.
.Pp
The
.Fn uuid_is_nil
function returns 1 if the UUID is nil-valued and 0 if it is not.
.Pp
The
.Fn uuid_to_string
function converts a UUID from binary representation to string representation.
Storage for the string is dynamically allocated and returned via the
.Fa str
argument.
This pointer should be passed to
.Xr free 3
to release the allocated storage when it is no longer needed.
.Pp
The
.Fn uuid_enc_le
and
.Fn uuid_enc_be
functions encode a binary representation of a UUID into an octet stream
in little-endian and big-endian byte-order, respectively.
The destination buffer must be pre-allocated by the caller, and must be
large enough to hold the 16-octet binary UUID.
.Pp
The
.Fn uuid_dec_le
and
.Fn uuid_dec_be
functions decode a UUID from an octet stream in little-endian and
big-endian byte-order, respectively.
.Sh RETURN VALUES
The
.Fn uuid_compare ,
.Fn uuid_create ,
.Fn uuid_create_nil ,
.Fn uuid_equal ,
.Fn uuid_from_string ,
.Fn uuid_hash ,
.Fn uuid_is_nil ,
and
.Fn uuid_to_string
functions return successful or unsuccessful completion status in the
.Fa status
argument.
Possible values are:
.Pp
.Bl -tag -width ".Dv uuid_s_invalid_string_uuid"
.It Dv uuid_s_ok
The function completed successfully.
.It Dv uuid_s_bad_version
The UUID does not have a known version.
.It Dv uuid_s_invalid_string_uuid
The string representation of a UUID is not valid.
.It Dv uuid_s_no_memory
Memory could not be allocated for the operation.
.El
.Sh SEE ALSO
.Xr uuidgen 1 ,
.Xr uuidgen 2
.Sh STANDARDS
The
.Fn uuid_compare ,
.Fn uuid_create ,
.Fn uuid_create_nil ,
.Fn uuid_equal ,
.Fn uuid_from_string ,
.Fn uuid_hash ,
.Fn uuid_is_nil ,
and
.Fn uuid_to_string
functions are compatible with the DCE 1.1 RPC specification.
.Sh HISTORY
The UUID functions first appeared in
.Nx 3.0 .


doc/CHANGES
LIST OF CHANGES FROM LAST RELEASE:			<$Revision: 1.1065 $>


[Note: This file does not mention every change made to the NetBSD source tree.
For an exhaustive list, please subscribe to the `source-changes' mailing list,
or see the mailing list archives.  For more information, send a message
containing just the word `help' to `majordomo@NetBSD.org'.]

------------------------------------------------------------------------------
PLEASE USE THE FOLLOWING STYLE WHEN ADDING ITEMS:

	theme: Content ...
		more content ...
		more content. [developer DATE-FORMAT]

That's it, one tab to start an item, and two tabs to indent the content.
Also please start the content with a capital and end it with a dot.

This is very important for the www scripts, otherwise it won't be able to
parse the file and the HTML output will be mangled.
------------------------------------------------------------------------------


Changes from NetBSD 4.0 to NetBSD 5.0:
	evbppc: Add support for Xilinx Virtex II-Pro/4-FX. [freza 20061202]
	mac68k: Add support for four more Nubus video cards, including the
		VillageTronic MacPicasso 340. [hauke 20061206]
	ixp425: Add support for NPE Ethernet. [scw 20061210]
	nfs: Remove NQNFS support from kernel. [yamt 20061227]
	mount_psshfs(8): Add puffs sshfs. [pooka 20061229]
	zaurus: Add support for Sharp C3x00 PDA. [ober 20070105]
	i386: Add support for Microsoft Xbox. [jmcneill 20070105]
	ntp: Import ntp 4.2.4 (20061228). Tracking of network interface
		address changes is now supported - no more ntpd
		restarts when interface addresses are re-configured
		(WLAN/DSL/ppp/...). [kardel 20070106]
	kernel: Global 'securelevel' variable removed; it is now internal to
		the secmodel. [elad 20070110]
	gcc(1): Updated to version 4.1.2-20070110. [mrg 20070111]
	acpitools: Added from FreeBSD via fvdl. [christos 20070114]
	scsictl(8): Added setspeed command, to set the highest speed that the
		optical drive should use for reading data,
		contributed by Kimura Fuyuki. [bouyer 20070122]
	i386, amd64: Add cmos(4), driver for x86 CMOS RAM. [dyoung 20070206]
	newlock2: Merge the newlock2 branch. Introduces a number of new
		kernel synchronization primitives, improves scalability
		on MP systems, and replaces the existing SA threads model
		with a 1:1 threading model. [ad 20070209]
	news68k: Implement generic softintr(9) support. [tsutsui 20070210]
	refuse: Added FUSE (file system in userspace) compatibility
		functionality, layered on top of libpuffs. [agc 20070211]
	i386: Added an opencrypto provider for VIA ACE (AES encryption
		instructions), ported from OpenBSD. [daniel 20070217]
	sdiff(1): Replaced by OpenBSD sdiff(1) (public domain). [rmind 20070219]
	btuart(4): Add support for Bluetooth HCI UART (H4) driver.
		[kiyohara 20070221]
	btuartd(8): Add support for Bluetooth HCI UART (H4) daemon.
		[kiyohara 20070221]
	mutex(9), rwlock(9): Support priority inheritance. [yamt 20070226]
	eeprom(8): Add the missing kernel bits on arm and powerpc that kept
		eeprom from actually changing firmware settings.
		[macallan 20070228]
	MAKEDEV(8), MAKEDEV.local(8): Let MAKEDEV.local use functions
		defined in MAKEDEV. [apb 20070303]
	x68k: Add generic softintr(9) support. [tsutsui 20070304]
	luna68k: Add generic softintr(9) support. [tsutsui 20070304]
	file(1): Update to 4.20.  [pooka 20070304]
	hfs: Add read-only Apple HFS+ file system support. [dillo 20070305]
	ipv6: Add ipv6 fast forward. [liamjfoy 20070307]
	openssh: Import 4.6. [christos 20070310]
	aiboost(4): ASUS AI Booster ACPI Hardware monitor driver,
		adapted from FreeBSD for envsys(4). [xtraeme 20070313]
	evbmips: Add Infineon ADM5120 port. [dyoung 20070320]
	i386, amd64: Intel(R) On Demand Clock Modulation driver, to save
		power and overheating in CPUs supporting the Thermal
		Monitor feature (TM). Adapted from OpenBSD/FreeBSD.
		See options(4) for more information. [xtraeme 20070320]
	kernel: Import of DRM (Direct Rendering Manager). Not yet enabled
		by default. [drochner 20070320]
	libc: Add support for LC_TIME. [manu 20070328]
	genfb(4): Generic framebuffer console driver with PCI and SBus
		frontends. [macallan 20070411]
	sgimips: Add support for SGI O2 PS/2 controller macekbc(4) and
		onboard display adapter crmfb(4). [jmcneill 20070412]
	sgimips: Import mavb(4) audio device driver for SGI O2 from
		OpenBSD. [jmcneill 20070412]
	sparc64: Add support for hardware monitoring on
		Ultra Enterprise 450. See envctrl(4). [tnn 20070414]
	puffs(4): Support NFS exporting puffs file servers. [pooka 20070416]
	xargs(1): Bring in xargs from FreeBSD to gain -I, -J etc. but keep our
		GNU compatible exit values. [christos 20070418]
	mount_9p(8): Add support to mount a 9P file service with puffs.
		[pooka 20070421]
	kernel: Remove the CHECK_ALT_xxx() defines and emul_find() used
		to locate files within an emulation's root directory.
		Instead get namei() to check the emulation root first.
		[dsl 20070422]
	prep: Add support for the MTX604 RAVEN based machine. Other RAVEN
		based machines are also likely to work now. [garbled 20070426]
	kernel: Added support for POSIX Asynchronous I/O. [rmind 20070430]
	puffs(3): Add puffs_framebuf(3), which abstracts the buffer
		management and event loop for networked file systems
		using puffs_cc(3). Convert mount_psshfs(8) and
		mount_9p(8) to use it. [pooka 20070505]
	kernel: Add the TFTPROOT kernel option for TFTP'ing root RAMdisk
		at root mount time. [manu 20070508]
	pf(4): Make 'nat' and 'rdr' translation rules obey state
		policy flags ('if-bound', 'group-bound', 'floating').
		Extend pf.conf(5) syntax to express the translation-state
		policies. [dyoung 20070510]
	idlelwp: yamt-idlelwp branch was merged.
		1. separate context switching and thread scheduling.
		2. implement idle lwp.
		3. clean up related MD/MI interfaces.
		4. make scheduler(s) modular.
		[yamt 20070518]
	ld.elf_so(1): Support Solaris $ORIGIN etc. expansions in paths.
		[christos 20070518]

usr.sbin/sdpd/uuid.c
/*	$NetBSD: uuid.c,v 1.1 2006/06/19 15:44:56 gdamore Exp $	*/

/*
 * uuid.c
 *
 * Copyright (c) 2005 Maksim Yevmenkin <m_evmenkin@yahoo.com>
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 * $Id: uuid.c,v 1.1 2006/06/19 15:44:56 gdamore Exp $
 * $FreeBSD: src/usr.sbin/bluetooth/sdpd/uuid.c,v 1.1 2005/01/05 18:37:37 emax Exp $
 */

#include <sys/cdefs.h>
__RCSID("$NetBSD: uuid.c,v 1.1 2006/06/19 15:44:56 gdamore Exp $");

#include <bluetooth.h>
#include <sdp.h>
#include <uuid.h>
#include "uuid-private.h"

uint128_t	uuid_base = {
	.b = {
		0x00, 0x00, 0x00, 0x00,
		0x00, 0x00,
		0x10, 0x00,
		0x80, 0x00,
		0x00, 0x80, 0x5f, 0x9b, 0x34, 0xfb
	}
};

uint128_t	uuid_public_browse_group = {
	.b  = {
		0x00, 0x00, 0x10, 0x02,
		0x00, 0x00,
		0x10, 0x00,
		0x80, 0x00,
		0x00, 0x80, 0x5f, 0x9b, 0x34, 0xfb
	}
};
